polynote-kernel/src/main/scala/polynote/kernel/lang/scal/ScalaSource.scala

```diff
@@ -1,29 +1,33 @@
 package polynote.kernel.lang.scal
 
 import cats.data.Ior
+import cats.effect.IO
 import cats.syntax.either._
 import polynote.kernel.EmptyCell
-import polynote.kernel.util.KernelReporter
+import polynote.kernel.lang.LanguageInterpreter
+import polynote.kernel.util.{CellContext, KernelContext, KernelReporter}
 import polynote.messages.CellID
 
 import scala.annotation.tailrec
 import scala.collection.immutable.{ListMap, Queue}
 import scala.reflect.internal.util.{ListOfNil, Position, RangePosition, SourceFile}
+import scala.tools.nsc.interactive.Global
 
 /**
   * Captures some Scala source in the context of a notebook, and provides compilation operations.
   *
   * TODO: the refined Interpreter type is really annoying; dictated by Global's path-dependent types. Can that be
   *       unwound?
   */
-class ScalaSource[Interpreter <: ScalaInterpreter](val interpreter: Interpreter)(
-  id: CellID,
-  availableSymbols: Set[Interpreter#Decl],
-  previousSources: List[ScalaSource[Interpreter]],
-  code: String
+class ScalaSource[G <: Global](
+  val global: G,
+  cellContext: CellContext,
+  previousSources: List[ScalaSource[G]],
+  notebookPackage: String,
+  afterParse: => Ior[Throwable, List[G#Tree]]
```

Comment:
Oh, and `ScalaSource` can now be constructed directly from trees - that's why this was moved to a parameter. The Scala parser is run in the constructor which takes `String` source code.